enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,445 @@
+will be, see: [Summary](#summary). I am thus leaving the title open-ended for
+further refinement once people review this enhancement proposal.
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+OVN-Kubernetes and openshift-sdn today support the feature egress IP on
+bare-metal platforms. Going forwards OpenShift will need to support the feature
+on conventional cloud provider platforms (such as: GCP, AWS, Azure). Cloud
+providers require additional management of the egress IP address via the cloud
+provider's cloud API. That is required since the cloud provider will not let the
+network plugin claim an additional IP on the node, which it does not recognize.
+
+To be able to do this we will need to add logic (to either a dedicated component
+or to the network plugins themselves) as to be able to handle this. This
+enhancement proposal concerns itself with a discussion surrounding questions
+such as:
+
+- If the component should be generic enough as to allow other resources the
+  capability to talk to the cloud API and modify host/cluster configurations? If
+  the answer is yes, then we probably need an separate enhancement proposal for
+  that, as this one only goes into the details of additional IP address
+  assignment (within the scope of egress IP)
+- If the component should be a dedicated OpenShift component / upstream one?
+- If something existing in Kubernetes/OpenShift already exists which satisfies
+  the requirements specified in this document? Certain resources have been found
+  in the initial investigation leading to this enhancement proposal, which seem
+  to indicate such. It is however not abundantly clear what those components do,
+  and what their limitations are, see: [1], [2], [3], [4], [5]
+
+Moreover, this document goes into detail concerning:
+
+- The cloud provider data model required as to manage egress IP assignments in
+  the cloud
+- The model and interaction between this future entity, the network plugin and
+  any additional component required for this interaction to happen.
+
+OpenShift 4.8 aims to support egress IP assignments on Azure and AWS. GCP is
+targeted for OpenShift 4.9. This document tries however to make an abstraction
+from the requirements and deadlines set for each cloud w.r.t. each OpenShift
+release. Hence it does not solely focus on AWS or Azure, but instead aims at
+defining an architecture which is cloud agnostic and which would be applied with
+minimal implementation efforts for additional clouds.
+
+[1]: https://kubernetes.io/docs/concepts/architecture/cloud-controller/
+[2]: https://github.com/kubernetes/cloud-provider
```

Comment:
cloud-provider API primarily targets load balancers today. It may be a good place (if we can build a consensus with sig-cloudprovider) to build some infrastructure around VPC configurations etc in an abstract API layer that decouples calling code from underlying cloud providers. But a layer of indirection also means it will suffer from some consensus latency.

Reply:
Yes, I agree. I will reach out to them on their k8s slack channel to see if there's any interest.

Reply:
I think there is some expectation that cloud-provider is only there to provide functionality needed by core kubernetes components. ie, if nothing in kubernetes/kubernetes itself is going to use the functionality, then it doesn't belong in cloud-provider

Reply:
OK, let's scrap the "maybe this can be an upstream kubernetes component?" or "included in an already existing kubernetes component?"